core/src/main/java/org/apache/iceberg/BaseTransaction.java

```diff
@@ -554,7 +554,7 @@ public TableMetadata refresh() {
     @Override
     @SuppressWarnings("ConsistentOverrides")
     public void commit(TableMetadata underlyingBase, TableMetadata metadata) {
-      if (underlyingBase != current) {
+      if (underlyingBase.metadataFileLocation().equals(current.metadataFileLocation())) {
```

Comment:
Should be 
```suggestion
      if (!underlyingBase.metadataFileLocation().equals(current.metadataFileLocation())) {
```

Reply:
On a higher level, I think this is the correct check although I guess it is possible that the response could contain a metadata.json path which is the same but the underlying response can be different in some other way. If that happens though It feels like something is much much more broken. 

Overall I think we probably should migrate to this check rather than object equality